src/emulator/hardware/builtin-gates/MipsAlu.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+const BuiltInGate = require('../BuiltInGate');
+
+/**
+ * Canonical truth table for the MipsAlu
+ */
+const TRUTH_TABLE = [
+
+  // Add
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b10, out: 0, cout: 1, set: 0 },
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 1, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 0, set: 0 },
+
+  // Sub
+  {a: 1, b: 1, na: 0, nb: 1, less: 0, cin: 1, op: 0b10, out: 0, cout: 1, set: 0 },
+
+  // And
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b00, out: 1, cout: 0, set: 0 },
+
+  // Or
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 0, cout: 0, set: 0 },
+
+];
+
+/**
+ * A 1-bit MipsAlu
+ */
+class MipsAlu extends BuiltInGate {
+  /**
+   * Logic:
+   *
+   *   a & b: op = 0
+   *   a | b: op = 1
+   *   a NOR b => ~a & ~b: op = 0, na = 1, nb = 1
+   *
+   * Math:
+   *
+   *   a + b: op = 2
+   *   a - b: op = 2, nb = 1, cin = 1
+   *   b - a: op = 2, na = 1, cin = 1
+   *
+   */
+  eval() {
+    const [ a, b, na, nb, less, cin, op ] = this.getInputPins().map(p => p.getValue());
+
+    const A = na ? +!a : a;
+    const B = nb ? +!b : b;
```

Comment:
This should be bitwise not `~`, not logical:

```js
const A = na ? ~a : a;
const B = nb ? ~b : b;
```

Reply:
I tried this first, but got `~1 === -2` and `~0 === ~1` which makes sense because the inputs are converted to 32 bit two's complement integers. But conceptually `a` and `b` are 1-bit so their negations should be `1 --> 0` and `0 --> 1` I thought?

( Values `A` and `B` here for example https://github.com/DmitrySoshnikov/hdl-js/blob/master/examples/__tests__/MipsAlu-hdl-test.js#L84 )

It turns out that the result `out` is the same for both ways, but the carry out is different:

```javascript
1 - 1 = 0
// Alu does it this way
a = 1, b = 1, cin = 1, nb = 1, na = 0

// If we use bitwise
A = a = 1
B = ~b = -2
sum = A + B + cin = 1 + -2 + 1 = 0
out = sum % 2 = 0
cout = sum / 2 = 0 // DIFFERENT

// If we use logical
A = a = 1
B = +!b = 0
sum = A + B + cin = 1 + 0 + 1 = 2
out = sum % 2 = 0
cout = sum / 2 = 1 // DIFFERENT
```
I thought the second carry out is expected behavior, that is why I used logical negation.

Reply:
Ah, in 1-bit [Not](https://github.com/DmitrySoshnikov/hdl-js/blob/master/src/emulator/hardware/builtin-gates/Not.js#L27) I actually just used `1 - a`. We can use the same to be unified with the actual `Not` implementation.